packages/datadog-plugin-fs/src/index.js

```diff
@@ -0,0 +1,490 @@
+    return makeFSTags(path, null, config, tracer)
+  }
+}
+
+function createFDTags (config, tracer) {
+  return function fdTags (fd) {
+    if (typeof this === 'object' && this !== null && this.fd) {
+      fd = this.fd
+    }
+    return makeFSTags(fd, null, config, tracer)
+  }
+}
+
+function createChownTags (config, tracer) {
+  return function chownTags (fd, uid, gid) {
+    const tags = makeFSTags(fd, null, config, tracer)
+    if (typeof uid === 'number') {
+      tags['file.uid'] = uid.toString()
+    }
+    if (typeof gid === 'number') {
+      tags['file.gid'] = gid.toString()
+    }
+    return tags
+  }
+}
+
+function createFchownTags (config, tracer) {
+  return function fchownTags (fd, uid, gid) {
+    if (typeof this === 'object' && this !== null && this.fd) {
+      gid = uid
+      uid = fd
+      fd = this.fd
+    }
+    const tags = makeFSTags(fd, null, config, tracer)
+    if (typeof uid === 'number') {
+      tags['file.uid'] = uid.toString()
+    }
+    if (typeof gid === 'number') {
+      tags['file.gid'] = gid.toString()
+    }
+    return tags
+  }
+}
+
+function getSymbolName (sym) {
+  return sym.description || sym.toString()
+}
+
+function createWrapCb (tracer, config, name, tagMaker) {
+  const makeTags = tagMaker(config, tracer)
+  name = 'fs.' + name
+  return function wrapFunction (fn) {
+    return tracer.wrap(name.toLowerCase(), function () {
+      if (typeof arguments[arguments.length - 1] !== 'function') {
+        return
+      }
+      const tags = makeTags.apply(this, arguments)
+      return tags ? { tags } : null
+    }, fn)
+  }
+}
+
+function createWrap (tracer, config, name, tagMaker) {
+  const makeTags = tagMaker(config, tracer)
+  name = 'fs.' + name
+  return function wrapSyncFunction (fn) {
+    return tracer.wrap(name.toLowerCase(), function () {
+      const tags = makeTags.apply(this, arguments)
+      return tags ? { tags } : null
+    }, fn)
+  }
+}
+
+function makeFSFlagTags (path, options, defaultFlag, config, tracer) {
+  const tags = makeFSTags(path, options, config, tracer)
+
+  if (tags) {
+    let flag = defaultFlag
+    if (typeof options === 'object' && options !== null) {
+      if (options.flag) {
+        flag = options.flag
+      } else if (options.flags) {
+        flag = options.flags
+      }
+    }
+    tags['file.flag'] = flag
+    return tags
+  }
+}
+
+function makeFSTags (path, options, config, tracer) {
+  path = options && 'fd' in options ? options.fd : path
+  const tags = {
+    'component': 'fs',
+    'service.name': config.service || `${tracer._service}-fs`
+  }
+
+  switch (typeof path) {
+    case 'object': {
+      if (path === null) return
+      const src = 'src' in path ? path.src : null
+      const dest = 'dest' in path ? path.dest : null
+      if (src || dest) {
+        tags['file.src'] = src.toString('utf8')
+        tags['file.dest'] = dest.toString('utf8')
+        tags['resource.name'] = (src || dest).toString('utf8')
```

Comment:
@brettlangdon Will this be problematic because of high cardinality? (applies to other spans as well)

Reply:
the cardinality would be really high for sure.

as long as we have tags for the src and destination people can always search/access that from App Analytics if they want to query based on a specific file.

we could try to quantize a bit by doing just the directory name?

Reply:
This would make it difficult to use especially if all files in the same folder for example. In general, I don't really expect this to be an issue since most of the usage for `fs` will come from outside a request, which I don't think we should trace. Also, if it works for tags, it should work for the resource name too. Maybe it's fine to leave it as is?

Reply:
The resource name is used as tags in metrics on the backend, so it must adhere to normal metric tags cardinality constraints.

What about serving static files? They could cause issues with cardinality.

Reply:
But then the whole point of the plugin is to provide visibility into what takes time to load. If we cannot provide that, what is the added value of the plugin?

Reply:
But it does. If it tells you an endpoint spent 50% of it's time loading files?

Reply:
So should we just not add a resource name at all?

Reply:
if you omit it'll likely default to operation name. Which is probably fine here.

Since you can see a breakdown of `fs` usage in your app, but then can use App Analytics if you actually care about doing breakdown based on the file path.

Reply:
cc @bengl

Reply:
I just pushed the change which omits the resource name, and tests that it defaults to the span name.